developer/appdev/articles/03-service-builder/03-creating-o-r-map/98-connecting-service-builder-to-an-external-database.markdown

```diff
@@ -0,0 +1,106 @@
+ 
+## Using a `DataSourceProvider` 
+
+Connecting to an external data base by creating and registering a `DataSourceProvider` as a JDK
+[ServiceProviderInterface](https://docs.oracle.com/javase/tutorial/sound/SPI-intro.html)
+(SPI) requires the fewest files and steps. 
+
+1.  Define your data source. One way is to use
+    [portal properties](/docs/7-2/deploy/-/knowledge_base/d/portal-properties)
+    in a `portal-ext.properties` file. Distinguish your data source from Liferay's default data source by giving it a prefix other
+than `jdbc.default.`. This example uses prefix `jdbc.ext.`. 
+
+```properties
+jdbc.ext.driverClassName=org.mariadb.jdbc.Driver
+jdbc.ext.password=userpassword
+jdbc.ext.url=jdbc:mariadb://localhost/external?useUnicode=true&characterEncoding=UTF-8&useFastDateParsing=false
+jdbc.ext.username=yourusername
+```
+
+2.  Restart your server if you defined your data source using portal properties.
+
+3.  Implement the
+    [`DataSourceProvider`](@platform-ref@/7.2-latest/javadocs/portal-kernel/com/liferay/portal/kernel/dao/jdbc/DataSourceProvider.html)
+    interface. Since the `DataSourceProvider` must be visible to your `*-service` module's class
+    loader, it's common to put the `DataSourceProvider` in that module. 
+
+    This example implementation uses
+    [`DataSourceFactoryUtil`](@platform-ref@/7.2-latest/javadocs/portal-kernel/com/liferay/portal/kernel/dao/jdbc/DataSourceFactoryUtil.html)
+    to create a data source from
+    [portal properties](/docs/7-2/deploy/-/knowledge_base/d/portal-properties)
+    that have the prefix `jdbc.ext.`. 
+
+```java
+import com.liferay.portal.kernel.dao.jdbc.DataSourceFactoryUtil;
+import com.liferay.portal.kernel.dao.jdbc.DataSourceProvider;
+import com.liferay.portal.kernel.util.PropsUtil;
+
+import javax.sql.DataSource;
+
+public class DataSourceProviderImpl implements DataSourceProvider {
+
+	@Override
+	public DataSource getDataSource() {
+		try {
+			return DataSourceFactoryUtil.initDataSource(
+				PropsUtil.getProperties("jdbc.ext.", true));
+		}
+		catch (Exception e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+}
+```
+
+4.  Register the implementation as a JDK
+    [ServiceProviderInterface](https://docs.oracle.com/javase/tutorial/sound/SPI-intro.html)
+    (SPI) in a
+    `/META-INF/services/com.liferay.portal.kernel.dao.jdbc.DataSourceProvider`
+    file in your `*-service` module. For example, this file registers the
+    `DataSourceProvider` implementation from the previous step.
+
+```
+com.liferay.external.data.source.test.internal.DataSourceProviderImpl
+````
+
+5.  [Run Service Builder](/docs/7-2/appdev/-/knowledge_base/a/running-service-builder). 
+
+6.  [Deploy](/docs/7-2/reference/-/knowledge_base/r/deploying-a-project)
+    your `-service` module. If your `DataSourceProvider` is in a 
+    different project, deploy it too. 
+
+Now your module's Service Builder services persist data to the external data
```

Comment:
This comes with a hidden cost.  Transactions between separate DataSources cannot be fully nested.  So depending on how the module with an external DataSource depends on other portal services rollbacks may not propagate between them.

Reply:
If nested transactions are a requirement than using an external DataSource may not be appropriate.

Reply:
@jhinkey we should give them this generic warning.  I don't want to go down into the details of what business logic can / cannot work - it's something they'll need to decide for themselves.  We can only give them a little warning so they know it's something to think about with their design / requirements.

Reply:
@Preston-Crary - How's a note like this ...

**Warning:** If your Service Builder services require nested transactions, 
using an external data source may not be appropriate for you. Transactions
between separate data sources cannot be fully nested. Rollbacks may not
propagate between a module that uses an external data source and @product@
services (or another app's services) that use a different data source.

Reply:
Perfect!